server/schedule/operator.go

```diff
@@ -14,88 +14,181 @@
+}
+
+// IsFinish checks if current step is finished.
+func (ap AddPeer) IsFinish(region *core.RegionInfo) bool {
+	if p := region.GetStorePeer(ap.ToStore); p != nil {
+		return region.GetPendingPeer(p.GetId()) == nil
 	}
-	return operatorStateToName[OperatorUnKnownState]
+	return false
+}
+
+// RemovePeer is an OperatorStep that removes a region peer.
+type RemovePeer struct {
+	FromStore uint64
 }
 
-// MarshalJSON returns the state as a JSON string
-func (o OperatorState) MarshalJSON() ([]byte, error) {
-	return []byte(strconv.Quote(o.String())), nil
+func (rp RemovePeer) String() string {
+	return fmt.Sprintf("remove peer on store %v", rp.FromStore)
 }
 
-// UnmarshalJSON  parses a JSON string into the OperatorState
-func (o *OperatorState) UnmarshalJSON(text []byte) error {
-	s, err := strconv.Unquote(string(text))
-	if err != nil {
-		return errors.Trace(err)
+// IsFinish checks if current step is finished.
+func (rp RemovePeer) IsFinish(region *core.RegionInfo) bool {
+	return region.GetStorePeer(rp.FromStore) == nil
+}
+
+// Operator contains execution steps generated by scheduler.
+type Operator struct {
+	desc        string
+	regionID    uint64
+	kind        core.ResourceKind
+	steps       []OperatorStep
+	currentStep int32
+	createTime  time.Time
+}
+
+// NewOperator creates a new operator.
+func NewOperator(desc string, regionID uint64, kind core.ResourceKind, steps ...OperatorStep) *Operator {
+	return &Operator{
+		desc:       desc,
+		regionID:   regionID,
+		kind:       kind,
+		steps:      steps,
+		createTime: time.Now(),
 	}
-	state, ok := operatorStateNameToValue[s]
-	if !ok {
-		*o = OperatorUnKnownState
-	} else {
-		*o = state
+}
+
+func (o *Operator) String() string {
+	return fmt.Sprintf("%s (kind:%s, region:%v, createAt:%s, currentStep:%v, steps:%v)", o.desc, o.kind, o.regionID, o.createTime, atomic.LoadInt32(&o.currentStep), o.steps)
+}
+
+// Desc returns the operator's short description.
+func (o *Operator) Desc() string {
+	return o.desc
+}
+
+// RegionID returns the region that operator is targeted.
+func (o *Operator) RegionID() uint64 {
+	return o.regionID
+}
+
+// ResourceKind returns operator's resource kind.
+func (o *Operator) ResourceKind() core.ResourceKind {
+	return o.kind
+}
+
+// ElapseTime returns duration since it was created.
+func (o *Operator) ElapseTime() time.Duration {
+	return time.Since(o.createTime)
+}
+
+// Len returns the operator's steps count.
+func (o *Operator) Len() int {
+	return len(o.steps)
+}
+
+// Step returns the i-th step.
+func (o *Operator) Step(i int) OperatorStep {
+	if i >= 0 && i < len(o.steps) {
+		return o.steps[i]
+	}
+	return nil
+}
+
+// Check checks if current step is finished, returns next step to take action.
+func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
+	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
+		if o.steps[int(step)].IsFinish(region) {
+			atomic.StoreInt32(&o.currentStep, step+1)
```

Comment:
`atomic.AddInt32(&o.currentStep, 1)`?

Reply:
If 2 goroutines call Check() at the same time, `currentStep` could be incorrectly set to a larger value.

Reply:
Maybe add some comments, seems not straight forward.